README.md
Nix Updater [![Build Status][BS img]][Build Status]
=================
[Build Status]: https://travis-ci.org/nixheads/nixupdater
[BS img]: https://api.travis-ci.org/nixheads/nixupdater.png

This is a tool that checks apt for updates and notify's the user accordingly.
It does this by updating the cache based on a timer the user can select in the config tool.


## License ![License](https://img.shields.io/badge/license-GPLv2-green.svg)

This project is under the GPLv2 license.

##Depends

Current depends required to run are:
- ``python-gtk2``
- ``python-apt``
- ``python-pyinotify``
- ``python-notify2``
- ``python-aptdaemon``

##Author
 - [Johnathan "ShaggyTwoDope" Jenkins](https://github.com/shaggytwodope/)


##Credits
- Brian Tomlinson - For ideas, suggestions, and generally helping me improve my code.
- AdrianKoshka - For helping word things for a sane person, and the great ideas.
- The Linux Lite users - For helping to report bugs or suggest features.



CHANGELOG
Changes for 1.0:
- Official stable release.


usr/share/lite-updater/lite-updater.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2016 Johnathan "Shaggytwodope" Jenkins <twodopeshaggy@gmail.com>
#
# Distributed under terms of the GPL2 license.

import shlex
import fcntl
from subprocess import Popen, PIPE
import pygtk
import gtk
from gtk import gdk
import glib
import apt
import time
import sys
import os
import gc
import pyinotify
import notify2
from threading import Thread
pygtk.require('2.0')
gtk.gdk.threads_init()


def run_check():
    global fl
    fl = open(os.path.realpath(__file__), 'r')
    try:
        fcntl.flock(fl, fcntl.LOCK_EX | fcntl.LOCK_NB)
    except:
        run_once_dialog()


def run_once_dialog():
    dialog = gtk.MessageDialog(None, 0, gtk.MESSAGE_WARNING, gtk.BUTTONS_OK,
                               appname + ' - Error')
    dialog.set_default_size(400, 250)
    dialog.format_secondary_text("There is another instance of " + appname +
                                 " already running.")
    response = dialog.run()
    if response == gtk.RESPONSE_OK:
        dialog.destroy()
        sys.exit()
    dialog.destroy()


def get_global_config():
    global config
    config = []
    try:
        for line in open(os.path.expanduser("~") + '/.config' +
                         '/lite-updater.conf'):
            line = line.rstrip()
            config.append(line)
    except:
        config = ["1800", "3600", "3000", "True"]
    return config


class Autostart(object):
    def user_autostart_path(self):
        config_autostart_path = os.getenv("HOME") + "/.config/autostart/"
        if not os.path.exists(config_autostart_path):
            os.makedirs(config_autostart_path)
        return config_autostart_path + "lite-updater.desktop"

    def is_installed(self):
        return os.path.exists(self.user_autostart_path())

    def install(self):
        with open(self.user_autostart_path(), "w") as desktop_file:
            desktop_file.write("""[Desktop Entry]
Type=Application
Version=0.1
Name=Lite-Updater
GenericName=Lite-Updater
Comment=Linux Lite Update Checker
Exec=/usr/bin/lite-updater
Terminal=false
Categories=GTK;Utility;
StartupNotify=true""")

        return True

    def uninstall(self):
        os.remove(self.user_autostart_path())


def checkcount():
    global upgrades
    cache = apt.Cache()
    cache.close()
    cache.open()
    upgrades = 0
    cache.upgrade(dist_upgrade=False)
    changes = cache.get_changes()
    if changes:
        counter = [change.name for change in changes]
        upgrades = (len(counter))
    return upgrades
    cache.close()
    gc.collect()


class Liteupdater:
    def __init__(self):
        self.config = get_global_config()
        self.icon = gtk.status_icon_new_from_file(inactive_icon)
        self.icon.set_tooltip("Idle")
        notify2.init("Lite-updater")
        self.icon.set_visible(True)
        self.update_running = False
        self.notifier = None
        self.watchdir = '/var/lib/apt'
        self.upgrades = 0
        self.icon.connect('activate', self.on_left_click)
        self.icon.connect('popup-menu', self.on_right_click)

    def update_cache(self, foo=None):
        self.set_state('working')
        cmd = "/usr/share/lite-updater/lite-cache-check.py"
        process = Popen(shlex.split(cmd), stdout=PIPE)
        process.communicate()
        self.update()
        gc.collect()

    def on_left_click(self, foo):
        cmd = "/usr/bin/gksu /usr/scripts/updates-gui-noprompt"
        process = Popen(shlex.split(cmd), stdout=PIPE)
        process.communicate()
        self.update()

    def checksources(self, foo):
        glib.spawn_async(["/usr/bin/gksu", "/usr/bin/software-properties-gtk"])

    def theconfig(self, foo):
        get_global_config()
        self.doconfig()

    def doconfig(self):
        cwindow = gtk.Window(gtk.WINDOW_TOPLEVEL)
        cwindow.set_position(gtk.WIN_POS_CENTER)
        cwindow.set_title("Updater Preferences")
        cwindow.set_icon_from_file(upgrade_icon)
        main_vbox = gtk.VBox(False, 5)
        main_vbox.set_border_width(10)
        cwindow.add(main_vbox)

        def default(self):
            config = ["1800", "3600", "3000", "True"]
            configfile = open(
                os.path.expanduser("~") + '/.config' + '/lite-updater.conf',
                "w")
            for item in config:
                configfile.write("%s\n" % item)
            spinner1.set_value(1)
            spinner2.set_value(30)
            spinner3.set_value(3)
            check.set_active(True)
            apply

        def apply(self):
            global config
            mins = spinner2.get_value_as_int()
            config[0] = mins * 60
            hours = spinner1.get_value_as_int()
            config[1] = hours * 60 * 60
            dseconds = spinner3.get_value_as_int()
            config[2] = dseconds * 1000
            config[3] = check.get_active()
            configfile = open(
                os.path.expanduser("~") + '/.config' + '/lite-updater.conf',
                "w")
            for item in config:
                configfile.write("%s\n" % item)
            return config

        def checksig(self):
            if check.get_active() is True:
                spinner3.set_sensitive(True)
                check.set_label('Notifications Enabled')
            elif check.get_active() is False:
                spinner3.set_sensitive(False)
                check.set_label('Notifications Disabled')

        def doauto_start(self):
            if Autostart().is_installed():
                pass
            else:
                Autostart().install()

        def doauto_remove(self):
            if Autostart().is_installed():
                Autostart().uninstall()
            else:
                pass

        def autostart(self):
            if Autostart().is_installed():
                doauto_remove(button)
                button.set_label("Enable Autostart")
            else:
                doauto_start(button)
                button.set_label("Disable Autostart")

        frame = gtk.Frame("Update Check Period")
        main_vbox.pack_start(frame, True, True, 0)
        vbox = gtk.VBox(False, 0)
        vbox.set_border_width(5)
        frame.add(vbox)
        hbox = gtk.HBox(False, 0)
        vbox.pack_start(hbox, True, True, 5)
        vbox2 = gtk.VBox(False, 0)
        hbox.pack_start(vbox2, True, True, 5)
        label = gtk.Label("Hours :")
        label.set_alignment(0, 0.5)
        vbox2.pack_start(label, False, True, 0)
        adj = gtk.Adjustment(0, 0, 23, 1, 10, 0)
        spinner1 = gtk.SpinButton(adj, 0, 0)
        policy = gtk.UPDATE_IF_VALID
        spinner1.set_update_policy(policy)
        spinner1.set_value(int(config[1]) / 60 / 60)
        spinner1.set_wrap(True)
        vbox2.pack_start(spinner1, False, True, 0)
        vbox2 = gtk.VBox(False, 0)
        hbox.pack_start(vbox2, True, True, 5)
        label = gtk.Label("Minutes :")
        label.set_alignment(0, 0.5)
        vbox2.pack_start(label, False, True, 0)
        adj = gtk.Adjustment(0, 5, 59, 1, 10, 0)
        policy = gtk.UPDATE_IF_VALID
        spinner2 = gtk.SpinButton(adj, 0, 0)
        spinner2.set_update_policy(policy)
        spinner2.set_value(int(config[0]) / 60)
        spinner2.set_wrap(True)
        vbox2.pack_start(spinner2, False, True, 0)
        vbox2 = gtk.VBox(False, 0)
        hbox.pack_start(vbox2, True, True, 5)
        frame = gtk.Frame("Notification Duration")
        main_vbox.pack_start(frame, True, True, 0)
        vbox = gtk.VBox(False, 0)
        vbox.set_border_width(5)
        frame.add(vbox)
        hbox = gtk.HBox(False, 0)
        vbox.pack_start(hbox, False, True, 5)
        vbox2 = gtk.VBox(False, 0)
        hbox.pack_start(vbox2, True, True, 5)
        label = gtk.Label("Seconds :")
        label.set_alignment(0, 0.5)
        vbox2.pack_start(label, False, True, 0)
        adj = gtk.Adjustment(0, 1, 60, 1, 10, 0)
        policy = gtk.UPDATE_IF_VALID
        spinner3 = gtk.SpinButton(adj, 0, 0)
        spinner3.set_update_policy(policy)
        spinner3.set_value(int(config[2]) / 1000)
        spinner3.set_wrap(True)
        spinner3.set_size_request(100, -1)
        vbox2.pack_start(spinner3, False, True, 0)
        vbox2 = gtk.VBox(False, 0)
        hbox.pack_start(vbox2, True, True, 5)
        check = gtk.CheckButton("Notifcations Enabled")
        if config[3] == "True":
            check.set_active(True)
            spinner3.set_sensitive(True)
        elif config[3] == "False":
            check.set_active(False)
            spinner3.set_sensitive(False)
            check.set_label('Notifications Disabled')
        check.connect("toggled", checksig)
        vbox.pack_start(check, True, True, 0)
        frame = gtk.Frame("Miscellaneous Settings")
        main_vbox.pack_start(frame, True, True, 0)
        vbox = gtk.VBox(False, 0)
        vbox.set_border_width(5)
        frame.add(vbox)
        hbox = gtk.HBox(False, 0)
        vbox.pack_start(hbox, False, True, 5)
        button = gtk.Button("Reset")
        button.connect("clicked", default)
        hbox.pack_start(button, True, True, 5)
        if Autostart().is_installed():
            button = gtk.Button("Disable Autostart")
        else:
            button = gtk.Button("Enable Autostart")
        button.connect("clicked", autostart)
        hbox.pack_start(button, True, True, 5)
        hbox = gtk.HBox(False, 0)
        main_vbox.pack_start(hbox, False, True, 0)
        buttonc = gtk.Button(stock=gtk.STOCK_CLOSE)
        buttonc.connect("clicked", lambda w: cwindow.destroy())
        hbox.pack_start(buttonc, True, True, 5)
        buttona = gtk.Button(stock=gtk.STOCK_APPLY)
        buttona.connect("clicked", apply)
        hbox.pack_start(buttona, True, True, 5)
        cwindow.show_all()

    def on_right_click(self, icon, button, time):
        menu = gtk.Menu()

        img1 = gtk.Image()
        img2 = gtk.Image()
        img3 = gtk.Image()

        about = gtk.ImageMenuItem('gtk-about', None)
        prefs = gtk.ImageMenuItem('gtk-preferences', None)
        updatechk = gtk.ImageMenuItem('Check for Updates')
        img1.set_from_file(working_icon)
        updatechk.set_image(img1)
        chsources = gtk.ImageMenuItem('Sources')
        img2.set_from_file(working_icon)
        chsources.set_image(img2)
        installupdate = gtk.ImageMenuItem('Update')
        img3.set_from_file(upgrade_icon)
        installupdate.set_image(img3)
        refresh = gtk.ImageMenuItem('gtk-refresh', None)
        refresh.set_label('Reload')
        quit = gtk.ImageMenuItem('gtk-quit', None)

        about.connect('activate', self.show_about_dialog)
        prefs.connect('activate', self.theconfig)
        sep2 = gtk.SeparatorMenuItem()
        updatechk.connect('activate', self.update_cache)
        chsources.connect('activate', self.checksources)
        installupdate.connect('activate', self.on_left_click)
        sep3 = gtk.SeparatorMenuItem()
        refresh.connect('activate', self.update)
        quit.connect('activate', self.quit)

        menu.append(about)
        menu.append(prefs)
        menu.append(sep2)
        menu.append(updatechk)
        menu.append(chsources)
        menu.append(installupdate)
        menu.append(sep3)
        menu.append(refresh)
        menu.append(quit)

        menu.show_all()
        menu.popup(None, None, gtk.status_icon_position_menu, button, time,
                   icon)

    def show_about_dialog(self, widget):
        about_dialog = gtk.AboutDialog()

        about_dialog.set_destroy_with_parent(True)
        about_dialog.set_name(appname)
        about_dialog.set_comments('A simple, lite update checker'
                                  ' for your tray.')
        about_dialog.set_website('https://www.linuxliteos.com')
        about_dialog.set_website_label('Website')
        about_dialog.set_icon(gdk.pixbuf_new_from_file(upgrade_icon))
        about_dialog.set_logo(gdk.pixbuf_new_from_file(logo))
        about_dialog.set_copyright('Copyright Linux Lite 2016')
        about_dialog.set_version(appver)
        about_dialog.set_authors(['Johnathan "ShaggyTwoDope" Jenkins'])
        about_dialog.set_wrap_license
        about_dialog.set_license(
            '''This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
MA 02110-1301, USA. ''')

        about_dialog.run()
        about_dialog.destroy()

    def quit(self, widget):
        self.notifier.stop()
        sys.exit(0)

    def set_state(self, state):
        self.tooltip_state = {
            'upgrade': 'Updates Available' + " " + str(self.upgrades),
            'inactive': 'No Updates Available',
            'working': 'Operations In Progress\nPlease wait...',
            'cache': 'Updating Package Cache\nPlease wait...'
        }

        self.upgrades = checkcount()
        icon = eval('{0}_icon'.format(state))
        self.icon.set_from_file(icon)
        self.icon.set_tooltip(self.tooltip_state[state])
        if config[3] == "True":
            n = notify2.Notification(appname, (self.tooltip_state[state]))
            nicon = gdk.pixbuf_new_from_file(icon)
            n.set_icon_from_pixbuf(nicon)
            n.set_timeout(int(config[2]))
            n.show()
        elif config[3] == "False":
            pass
        gc.collect()

    def update(self, ev=None):
        if self.update_running is False:
            self.update_running = True
            self.set_state('working')
            if self.upgrades:
                state = 'upgrade'
            else:
                state = 'inactive'

            self.set_state(state)
        self.update_running = False
        gc.collect()

    def timer_update(self):
        while True:
            self.update_cache()
            gc.collect()
            time.sleep(int(config[0] + config[1]))

    def main(self):
        wm = pyinotify.WatchManager()
        mask = pyinotify.IN_DELETE | pyinotify.IN_CREATE | pyinotify.IN_MODIFY
        self.notifier = pyinotify.ThreadedNotifier(wm, self.update)
        wm.add_watch(self.watchdir, mask)
        self.notifier.start()

        t = Thread(target=self.timer_update)
        t.daemon = True
        t.start()
        gc.collect()
        gtk.main()


if __name__ == "__main__":
    appname = 'Lite Updater'
    appver = '1.0-0100'
    iconpath = '/usr/share/lite-updater/icons/'
    inactive_icon = iconpath + 'updates-idle.png'
    working_icon = iconpath + 'aptdaemon-working.png'
    upgrade_icon = iconpath + 'updates-available.png'
    logo = iconpath + 'lite_logo.png'
    fl = 0
    app = Liteupdater()
    try:
        run_check()
        app.main()
    except KeyboardInterrupt:
        app.quit()


usr/share/nixupdater/nix-cache-check.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2016 Johnathan "Shaggytwodope" Jenkins <twodopeshaggy@gmail.com>
#
# Distributed under terms of the GPL2 license.

import sys
from gi.repository import GLib
from aptdaemon.client import AptClient

loop = GLib.MainLoop()
aptclient = AptClient()


def on_finished_update(trans, exit):
    if exit == "exit-success":
        print("successful cache update")
        loop.quit()
        sys.exit(0)
    else:
        print("error updating cache")
        sys.exit(1)
    return True


def do_update():
    trans_update = aptclient.update_cache()
    trans_update.connect("finished", on_finished_update)
    trans_update.run()
    return False


if __name__ == "__main__":
    try:
        GLib.timeout_add(50, do_update)
        loop.run()
    except:
        print("something unknown went wrong")
        sys.exit(1)


usr/share/lite-updater/lite-cache-check.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2016 Johnathan "Shaggytwodope" Jenkins <twodopeshaggy@gmail.com>
#
# Distributed under terms of the GPL2 license.

import sys
from gi.repository import GLib
from aptdaemon.client import AptClient

loop = GLib.MainLoop()
aptclient = AptClient()


def on_finished_update(trans, exit):
    if exit == "exit-success":
        print("successful cache update")
        loop.quit()
        sys.exit(0)
    else:
        print("error updating cache")
        sys.exit(1)
    return True


def do_update():
    trans_update = aptclient.update_cache()
    trans_update.connect("finished", on_finished_update)
    trans_update.run()
    return False


if __name__ == "__main__":
    try:
        GLib.timeout_add(50, do_update)
        loop.run()
    except:
        print("something unknown went wrong")
        sys.exit(1)


usr/share/nixupdater/nixupdater.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2016 Johnathan "Shaggytwodope" Jenkins <twodopeshaggy@gmail.com>
#
# Distributed under terms of the GPL2 license.

import shlex
import fcntl
from subprocess import Popen, PIPE
import pygtk
import gtk
from gtk import gdk
import glib
import apt
import time
import sys
import os
import gc
import pyinotify
import notify2
from threading import Thread
pygtk.require('2.0')
gtk.gdk.threads_init()


def run_check():
    global fl
    fl = open(os.path.realpath(__file__), 'r')
    try:
        fcntl.flock(fl, fcntl.LOCK_EX | fcntl.LOCK_NB)
    except:
        run_once_dialog()


def run_once_dialog():
    dialog = gtk.MessageDialog(None, 0, gtk.MESSAGE_WARNING, gtk.BUTTONS_OK,
                               appname + ' - Error')
    dialog.set_default_size(400, 250)
    dialog.format_secondary_text("There is another instance of " + appname +
                                 " already running.")
    response = dialog.run()
    if response == gtk.RESPONSE_OK:
        dialog.destroy()
        sys.exit()
    dialog.destroy()


def get_global_config():
    global config
    config = []
    try:
        for line in open(os.path.expanduser("~") + '/.config' +
                         '/nixupdater.conf'):
            line = line.rstrip()
            config.append(line)
    except:
        config = ["1800", "3600", "3000", "True"]
    return config


class Autostart(object):
    def user_autostart_path(self):
        config_autostart_path = os.getenv("HOME") + "/.config/autostart/"
        if not os.path.exists(config_autostart_path):
            os.makedirs(config_autostart_path)
        return config_autostart_path + "nixupdater.desktop"

    def is_installed(self):
        return os.path.exists(self.user_autostart_path())

    def install(self):
        with open(self.user_autostart_path(), "w") as desktop_file:
            desktop_file.write("""[Desktop Entry]
Type=Application
Version=1.0
Name=Nix Updater
GenericName=Nix Updater
Comment=Linux Update Checker
Exec=/usr/bin/nixupdater
Terminal=false
Categories=GTK;Utility;
StartupNotify=true""")

        return True

    def uninstall(self):
        os.remove(self.user_autostart_path())


def checkcount():
    global upgrades
    cache = apt.Cache()
    cache.close()
    cache.open()
    upgrades = 0
    cache.upgrade(dist_upgrade=False)
    changes = cache.get_changes()
    if changes:
        counter = [change.name for change in changes]
        upgrades = (len(counter))
    return upgrades
    cache.close()
    gc.collect()


class Nixupdater:
    def __init__(self):
        self.config = get_global_config()
        self.icon = gtk.status_icon_new_from_file(inactive_icon)
        self.icon.set_tooltip("Idle")
        notify2.init("Nix Updater")
        self.icon.set_visible(True)
        self.update_running = False
        self.notifier = None
        self.watchdir = '/var/lib/apt'
        self.upgrades = 0
        self.icon.connect('activate', self.on_left_click)
        self.icon.connect('popup-menu', self.on_right_click)

    def update_cache(self, foo=None):
        self.set_state('working')
        cmd = "/usr/share/nixupdater/nix-cache-check.py"
        process = Popen(shlex.split(cmd), stdout=PIPE)
        process.communicate()
        self.update()
        gc.collect()

    def on_left_click(self, foo):
        cmd = "/usr/bin/gksu /usr/scripts/updates-gui-noprompt"
        process = Popen(shlex.split(cmd), stdout=PIPE)
        process.communicate()
        self.update()

    def checksources(self, foo):
        glib.spawn_async(["/usr/bin/gksu", "/usr/bin/software-properties-gtk"])

    def theconfig(self, foo):
        get_global_config()
        self.doconfig()

    def doconfig(self):
        cwindow = gtk.Window(gtk.WINDOW_TOPLEVEL)
        cwindow.set_position(gtk.WIN_POS_CENTER)
        cwindow.set_title("Updater Preferences")
        cwindow.set_icon_from_file(upgrade_icon)
        main_vbox = gtk.VBox(False, 5)
        main_vbox.set_border_width(10)
        cwindow.add(main_vbox)

        def default(self):
            config = ["1800", "3600", "3000", "True"]
            configfile = open(
                os.path.expanduser("~") + '/.config' + '/nixupdater.conf',
                "w")
            for item in config:
                configfile.write("%s\n" % item)
            spinner1.set_value(1)
            spinner2.set_value(30)
            spinner3.set_value(3)
            check.set_active(True)
            apply

        def apply(self):
            global config
            mins = spinner2.get_value_as_int()
            config[0] = mins * 60
            hours = spinner1.get_value_as_int()
            config[1] = hours * 60 * 60
            dseconds = spinner3.get_value_as_int()
            config[2] = dseconds * 1000
            config[3] = check.get_active()
            configfile = open(
                os.path.expanduser("~") + '/.config' + '/nixupdater.conf',
                "w")
            for item in config:
                configfile.write("%s\n" % item)
            return config

        def checksig(self):
            if check.get_active() is True:
                spinner3.set_sensitive(True)
                check.set_label('Notifications Enabled')
            elif check.get_active() is False:
                spinner3.set_sensitive(False)
                check.set_label('Notifications Disabled')

        def doauto_start(self):
            if Autostart().is_installed():
                pass
            else:
                Autostart().install()

        def doauto_remove(self):
            if Autostart().is_installed():
                Autostart().uninstall()
            else:
                pass

        def autostart(self):
            if Autostart().is_installed():
                doauto_remove(button)
                button.set_label("Enable Autostart")
            else:
                doauto_start(button)
                button.set_label("Disable Autostart")

        frame = gtk.Frame("Update Check Period")
        main_vbox.pack_start(frame, True, True, 0)
        vbox = gtk.VBox(False, 0)
        vbox.set_border_width(5)
        frame.add(vbox)
        hbox = gtk.HBox(False, 0)
        vbox.pack_start(hbox, True, True, 5)
        vbox2 = gtk.VBox(False, 0)
        hbox.pack_start(vbox2, True, True, 5)
        label = gtk.Label("Hours :")
        label.set_alignment(0, 0.5)
        vbox2.pack_start(label, False, True, 0)
        adj = gtk.Adjustment(0, 0, 23, 1, 10, 0)
        spinner1 = gtk.SpinButton(adj, 0, 0)
        policy = gtk.UPDATE_IF_VALID
        spinner1.set_update_policy(policy)
        spinner1.set_value(int(config[1]) / 60 / 60)
        spinner1.set_wrap(True)
        vbox2.pack_start(spinner1, False, True, 0)
        vbox2 = gtk.VBox(False, 0)
        hbox.pack_start(vbox2, True, True, 5)
        label = gtk.Label("Minutes :")
        label.set_alignment(0, 0.5)
        vbox2.pack_start(label, False, True, 0)
        adj = gtk.Adjustment(0, 5, 59, 1, 10, 0)
        policy = gtk.UPDATE_IF_VALID
        spinner2 = gtk.SpinButton(adj, 0, 0)
        spinner2.set_update_policy(policy)
        spinner2.set_value(int(config[0]) / 60)
        spinner2.set_wrap(True)
        vbox2.pack_start(spinner2, False, True, 0)
        vbox2 = gtk.VBox(False, 0)
        hbox.pack_start(vbox2, True, True, 5)
        frame = gtk.Frame("Notification Duration")
        main_vbox.pack_start(frame, True, True, 0)
        vbox = gtk.VBox(False, 0)
        vbox.set_border_width(5)
        frame.add(vbox)
        hbox = gtk.HBox(False, 0)
        vbox.pack_start(hbox, False, True, 5)
        vbox2 = gtk.VBox(False, 0)
        hbox.pack_start(vbox2, True, True, 5)
        label = gtk.Label("Seconds :")
        label.set_alignment(0, 0.5)
        vbox2.pack_start(label, False, True, 0)
        adj = gtk.Adjustment(0, 1, 60, 1, 10, 0)
        policy = gtk.UPDATE_IF_VALID
        spinner3 = gtk.SpinButton(adj, 0, 0)
        spinner3.set_update_policy(policy)
        spinner3.set_value(int(config[2]) / 1000)
        spinner3.set_wrap(True)
        spinner3.set_size_request(100, -1)
        vbox2.pack_start(spinner3, False, True, 0)
        vbox2 = gtk.VBox(False, 0)
        hbox.pack_start(vbox2, True, True, 5)
        check = gtk.CheckButton("Notifcations Enabled")
        if config[3] == "True":
            check.set_active(True)
            spinner3.set_sensitive(True)
        elif config[3] == "False":
            check.set_active(False)
            spinner3.set_sensitive(False)
            check.set_label('Notifications Disabled')
        check.connect("toggled", checksig)
        vbox.pack_start(check, True, True, 0)
        frame = gtk.Frame("Miscellaneous Settings")
        main_vbox.pack_start(frame, True, True, 0)
        vbox = gtk.VBox(False, 0)
        vbox.set_border_width(5)
        frame.add(vbox)
        hbox = gtk.HBox(False, 0)
        vbox.pack_start(hbox, False, True, 5)
        button = gtk.Button("Reset")
        button.connect("clicked", default)
        hbox.pack_start(button, True, True, 5)
        if Autostart().is_installed():
            button = gtk.Button("Disable Autostart")
        else:
            button = gtk.Button("Enable Autostart")
        button.connect("clicked", autostart)
        hbox.pack_start(button, True, True, 5)
        hbox = gtk.HBox(False, 0)
        main_vbox.pack_start(hbox, False, True, 0)
        buttonc = gtk.Button(stock=gtk.STOCK_CLOSE)
        buttonc.connect("clicked", lambda w: cwindow.destroy())
        hbox.pack_start(buttonc, True, True, 5)
        buttona = gtk.Button(stock=gtk.STOCK_APPLY)
        buttona.connect("clicked", apply)
        hbox.pack_start(buttona, True, True, 5)
        cwindow.show_all()

    def on_right_click(self, icon, button, time):
        menu = gtk.Menu()

        img1 = gtk.Image()
        img2 = gtk.Image()
        img3 = gtk.Image()

        about = gtk.ImageMenuItem('gtk-about', None)
        prefs = gtk.ImageMenuItem('gtk-preferences', None)
        updatechk = gtk.ImageMenuItem('Check for Updates')
        img1.set_from_file(working_icon)
        updatechk.set_image(img1)
        chsources = gtk.ImageMenuItem('Sources')
        img2.set_from_file(working_icon)
        chsources.set_image(img2)
        installupdate = gtk.ImageMenuItem('Update')
        img3.set_from_file(upgrade_icon)
        installupdate.set_image(img3)
        refresh = gtk.ImageMenuItem('gtk-refresh', None)
        refresh.set_label('Reload')
        quit = gtk.ImageMenuItem('gtk-quit', None)

        about.connect('activate', self.show_about_dialog)
        prefs.connect('activate', self.theconfig)
        sep2 = gtk.SeparatorMenuItem()
        updatechk.connect('activate', self.update_cache)
        chsources.connect('activate', self.checksources)
        installupdate.connect('activate', self.on_left_click)
        sep3 = gtk.SeparatorMenuItem()
        refresh.connect('activate', self.update)
        quit.connect('activate', self.quit)

        menu.append(about)
        menu.append(prefs)
        menu.append(sep2)
        menu.append(updatechk)
        menu.append(chsources)
        menu.append(installupdate)
        menu.append(sep3)
        menu.append(refresh)
        menu.append(quit)

        menu.show_all()
        menu.popup(None, None, gtk.status_icon_position_menu, button, time,
                   icon)

    def show_about_dialog(self, widget):
        about_dialog = gtk.AboutDialog()

        about_dialog.set_destroy_with_parent(True)
        about_dialog.set_name(appname)
        about_dialog.set_comments('A simple, lite update checker'
                                  ' for your tray.')
        about_dialog.set_website('https://github.com/nixheads/nixupdater')
        about_dialog.set_website_label('Website')
        about_dialog.set_icon(gdk.pixbuf_new_from_file(upgrade_icon))
        about_dialog.set_logo(gdk.pixbuf_new_from_file(logo))
        about_dialog.set_copyright('Copyright Johnathan Jenkins 2016')
        about_dialog.set_version(appver)
        about_dialog.set_authors(['Johnathan "ShaggyTwoDope" Jenkins'])
        about_dialog.set_wrap_license
        about_dialog.set_license(
            '''This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
MA 02110-1301, USA. ''')

        about_dialog.run()
        about_dialog.destroy()

    def quit(self, widget):
        self.notifier.stop()
        sys.exit(0)

    def set_state(self, state):
        self.tooltip_state = {
            'upgrade': 'Updates Available' + " " + str(self.upgrades),
            'inactive': 'No Updates Available',
            'working': 'Operations In Progress\nPlease wait...',
            'cache': 'Updating Package Cache\nPlease wait...'
        }

        self.upgrades = checkcount()
        icon = eval('{0}_icon'.format(state))
        self.icon.set_from_file(icon)
        self.icon.set_tooltip(self.tooltip_state[state])
        if config[3] == "True":
            n = notify2.Notification(appname, (self.tooltip_state[state]))
            nicon = gdk.pixbuf_new_from_file(icon)
            n.set_icon_from_pixbuf(nicon)
            n.set_timeout(int(config[2]))
            n.show()
        elif config[3] == "False":
            pass
        gc.collect()

    def update(self, ev=None):
        if self.update_running is False:
            self.update_running = True
            self.set_state('working')
            if self.upgrades:
                state = 'upgrade'
            else:
                state = 'inactive'

            self.set_state(state)
        self.update_running = False
        gc.collect()

    def timer_update(self):
        while True:
            self.update_cache()
            gc.collect()
            time.sleep(int(config[0] + config[1]))

    def main(self):
        wm = pyinotify.WatchManager()
        mask = pyinotify.IN_DELETE | pyinotify.IN_CREATE | pyinotify.IN_MODIFY
        self.notifier = pyinotify.ThreadedNotifier(wm, self.update)
        wm.add_watch(self.watchdir, mask)
        self.notifier.start()

        t = Thread(target=self.timer_update)
        t.daemon = True
        t.start()
        gc.collect()
        gtk.main()


if __name__ == "__main__":
    appname = 'Nix Updater'
    appver = '1.0.0'
    iconpath = '/usr/share/nixupdater/icons/'
    inactive_icon = iconpath + 'updates-idle.png'
    working_icon = iconpath + 'aptdaemon-working.png'
    upgrade_icon = iconpath + 'updates-available.png'
    logo = iconpath + 'nix_logo.png'
    fl = 0
    app = Nixupdater()
    try:
        run_check()
        app.main()
    except KeyboardInterrupt:
        app.quit()
